dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -282,53 +284,105 @@ SegmentPtr Segment::newSegment( //
         context.storage_pool->newMetaPageId());
 }
 
-SegmentPtr Segment::restoreSegment( //
-    const LoggerPtr & parent_log,
-    DMContext & context,
-    PageIdU64 segment_id)
+inline void readSegmentMetaInfo(ReadBuffer & buf, Segment::SegmentMetaInfo & segment_info)
 {
-    Page page = context.storage_pool->metaReader()->read(segment_id); // not limit restore
+    readIntBinary(segment_info.version, buf);
+    readIntBinary(segment_info.epoch, buf);
 
-    ReadBufferFromMemory buf(page.data.begin(), page.data.size());
-    SegmentFormat::Version version;
-
-    readIntBinary(version, buf);
-    UInt64 epoch;
-    RowKeyRange rowkey_range;
-    PageIdU64 next_segment_id, delta_id, stable_id;
-
-    readIntBinary(epoch, buf);
-
-    switch (version)
+    switch (segment_info.version)
     {
     case SegmentFormat::V1:
     {
         HandleRange range;
         readIntBinary(range.start, buf);
         readIntBinary(range.end, buf);
-        rowkey_range = RowKeyRange::fromHandleRange(range);
+        segment_info.range = RowKeyRange::fromHandleRange(range);
         break;
     }
     case SegmentFormat::V2:
     {
-        rowkey_range = RowKeyRange::deserialize(buf);
+        segment_info.range = RowKeyRange::deserialize(buf);
         break;
     }
     default:
-        throw Exception(fmt::format("Illegal version: {}", version), ErrorCodes::LOGICAL_ERROR);
+        throw Exception(fmt::format("Illegal version: {}", segment_info.version), ErrorCodes::LOGICAL_ERROR);
     }
 
-    readIntBinary(next_segment_id, buf);
-    readIntBinary(delta_id, buf);
-    readIntBinary(stable_id, buf);
+    readIntBinary(segment_info.next_segment_id, buf);
+    readIntBinary(segment_info.delta_id, buf);
+    readIntBinary(segment_info.stable_id, buf);
+}
 
-    auto delta = DeltaValueSpace::restore(context, rowkey_range, delta_id);
-    auto stable = StableValueSpace::restore(context, stable_id);
-    auto segment = std::make_shared<Segment>(parent_log, epoch, rowkey_range, segment_id, next_segment_id, delta, stable);
+SegmentPtr Segment::restoreSegment( //
+    const LoggerPtr & parent_log,
+    DMContext & context,
+    PageIdU64 segment_id)
+{
+    Page page = context.storage_pool->metaReader()->read(segment_id); // not limit restore
+
+    ReadBufferFromMemory buf(page.data.begin(), page.data.size());
+    Segment::SegmentMetaInfo segment_info;
+    readSegmentMetaInfo(buf, segment_info);
+
+    auto delta = DeltaValueSpace::restore(context, segment_info.range, segment_info.delta_id);
+    auto stable = StableValueSpace::restore(context, segment_info.stable_id);
+    auto segment = std::make_shared<Segment>(parent_log, segment_info.epoch, segment_info.range, segment_id, segment_info.next_segment_id, delta, stable);
 
     return segment;
 }
 
+Segment::SegmentMetaInfos Segment::readAllSegmentsMetaInfoInRange( //
```

Comment:
For each region, we need to iterate all segments' metainfo. I think it is not optimized. I think we cache the `range_end -> segment_id` in `temp_ps ` in memory. And then only need to iterate from the first useful segment.

Reply:
I think we cat do the cache by `NamespaceId`. E.g. `std::unordered_map< NamespaceId, std::map<RowKeyValueRef, SegmentID>>`.

Reply:
OK, I will add it.